### PR TITLE
WX: Handle quadratic curves

### PR DESCRIPTION
**Pull request by madig**
Fixes https://github.com/trufont/trufont/issues/598.

http://unifiedfontobject.org/versions/ufo3/glyphs/glif/#contour says a contour only containing off-curves must be treated as quadratic. Still need to implement that.

Also: more resilience to invalid curves like 3 off-curves and a `curve`? Ideally, this is caught and the UI displays the glyph as containing invalid contours, allowing the user to go in and fix it...

### CODE REVIEW

**Comment by madig** [comment]
Not sure my handling of all-off-curve contours is clean?

**Comment by madig** [comment]
Also, I'd love to have a unit test for this :thinking:

**Comment by adrientetar** [comment]
Thanks, I have exams this week but I'll look at it in the we

**Comment by madig** [comment]
With https://github.com/trufont/tfont/pull/8 I can open the glyphs and they look pretty crappy. Mixture of quadratic and cubic curves in e.g. `B`.

**Comment by madig** [comment]
Oops, because I don't handle quadratic off-curves between two non-smooth points, they get misdetected as cubics.

File: src/trufont/objects/factories.py [hunk 1]
```diff
@@ -96,7 +96,36 @@ def pathGraphicsPathFactory(path):
             stack.append(point.y)
             if point.type == "curve":
                 graphicsPath.AddCurveToPoint(*stack)
-                stack = []
+                stack.clear()
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: stack = [] does the same, but does not need a method lookup.

File: src/trufont/objects/factories.py [hunk 2]
```diff
@@ -96,7 +96,36 @@ def pathGraphicsPathFactory(path):
...
+            #     path.
+            elif point.type == "qcurve" or point is points[-1]:
+                if len(stack) == 1:
+                    graphicsPath.AddQuadCurveToPoint(*stack)
+                    stack.clear()
+                else:
+                    stack_offcurves = trufont.util.misc.grouper(stack, 2)
+                    for (cx1, cy1), (cx2, cy2) in trufont.util.misc.pairwise(
+                        stack_offcurves
+                    ):
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: - This assumes number of points is even, right?
- That's a lot of intermediary functions for a simple iteration, we can instead do something like:

```python
offset = 0
size = len(stack)
while 4 + offset < size:
  points = stack[offset:4+offset]
  ...
  offset += 2
```

Can be rewritten in terms of for loop as e.g.:

```python
for offset in range(0, len(stack) - 4, 2):
  points = stack[offset:4+offset]
```

**Comment by adrientetar** [comment]
- How common are "all off curves" paths? Which tools produce this and which file formats and editors support it?
- This will need some changes in tfont first,
  * add support for qcurve in segment-specific ops of tfont, in path.py. search for pts_len
  * in tfont slice.py, the slicing code depends on the presence of segments, and segments are formed by the presence of an on curve point.  
just an example of code that "all off curves" paths will break in tfont and trufont
- I'd rather we keep "all off curves" out at least for now

**Comment by madig** [comment]
> How common are "all off curves" paths? Which tools produce this and which file formats and editors support it?

Dunno, but the UFO spec talks about it. Can do a `if all(offcurve): explode()` for now? Speaking of explosions, I'd like TruFont to handle exceptions while rendering more gracefully than now, where it basically corrupts the UI and probably the entire application. Todo.

**Comment by madig** [comment]
I just realized that the `len(stack) == 1` line should actually be `== 4`. D'oh.

**Comment by madig** [comment]
This actually renders glyphs correctly now. Thanks for the algorithm!

Now I need to look at handles being drawn without a line if we're looking at qcurves. Anything else?

I can't seem to find a way to properly test the paths though :(

**Comment by madig** [comment]
Heh.
```
  File "/home/nikolaus/Entwicklung/tfont/src/tfont/objects/path.py", line 281, in bounds
    "cannot compute bounds for %r segment" % self.type)
NotImplementedError: cannot compute bounds for 'qcurve' segment
```

File: src/trufont/objects/factories.py [hunk 3]
```diff
@@ -76,13 +76,16 @@ def pathGraphicsPathFactory(path):
     points = path._points
     if not points:
         return graphicsPath
+    if all(point.type is None for point in points):
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: This is a hot code path. Why make a linear time assertion for a possibility that we don't know is possible?

File: src/trufont/objects/factories.py [hunk 4]
```diff
@@ -76,13 +76,16 @@ def pathGraphicsPathFactory(path):
...
+        # https://github.com/trufont/trufont/pull/599#issuecomment-441391756
+        raise ValueError("All-off-curve paths are not currently supported.")
+
     start = points[0]
     open_ = skip = start.type == "move"
     if open_:
         graphicsPath.MoveToPoint(start.x, start.y)
     else:
         start = points[-1]
-        assert start.type is not None
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: Can you bring this back, a path is a suite of segments and a segment ends with an oncurve.

File: src/trufont/objects/factories.py [hunk 5]
```diff
@@ -97,6 +100,36 @@ def pathGraphicsPathFactory(path):
             if point.type == "curve":
                 graphicsPath.AddCurveToPoint(*stack)
                 stack = []
+            elif point.type == "qcurve" or (not open_ and point is points[-1]):
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: what's (not open_ and point is points[-1]) ? what should the last point be assumed to be qcurve?

File: src/trufont/objects/factories.py [hunk 6]
```diff
@@ -97,6 +100,36 @@ def pathGraphicsPathFactory(path):
...
+                assert len(stack) % 2 == 0, "Stack must have even number of points."
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: Comment is basically the same thing as the code before, so unneeded. If the assert fails, the LOC will be included in the traceback either way.

File: src/trufont/objects/factories.py [hunk 7]
```diff
@@ -97,6 +100,36 @@ def pathGraphicsPathFactory(path):
...
+                #     looking at the qcurve: stack == [cx, cy, qx, qy], ready to
+                #     AddQuadCurveToPoint.
+                #  2. We have more than one off-curve on the stack and are looking at
+                #     the last element, the qcurve. Here, TrueType's implied on-curve
+                #     principle applies that requires that an implied on-curve point
+                #     is inserted halfway between all off-curve points. The last
+                #     off-curve is then used as the control-point of the qcurve.
+                #  3. The stack is full of off-curves and we have reached the last
+                #     point. An all off-curve path is to be handled like a quadratic
+                #     path.
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: This is far too long and I think unneeded

File: src/trufont/util/drawing.py [hunk 8]
```diff
@@ -440,9 +440,14 @@ def drawLayerPoints(
             # others
             prev = points[-1]
             breakHandle = path.open
+            quadraticPath = any(point.type == "qcurve" for point in points)
```
Comment (adrientetar) [inline src/trufont/util/drawing.py]: ugh. this whole code is written to read the path only once. why is this needed?

File: src/trufont/util/drawing.py [hunk 9]
```diff
@@ -440,9 +440,14 @@ def drawLayerPoints(
...
             for point in points:
                 isOff = point.type is None
-                if not breakHandle and isOff != (prev.type is None):
+                if (
+                    not breakHandle
+                    and isOff != (prev.type is None)
+                    and not quadraticPath
+                ):
```
Comment (adrientetar) [inline src/trufont/util/drawing.py]: do we even need to change the code that works for curve?

**Comment by adrientetar** [comment]
> Dunno, but the UFO spec talks about it.

If you don't know, then don't bother. There's far enough work with things that actually happen, and every new LOC in the codebase is extra maintenance and cognitive burden for the next person working with that code. And they'll be misled into thinking that it's something that's actually expected.

UFO spec tends to go and "support every possibility" but it's more harmful than not. It weakens invariants about the data and makes code more complicated than need be. If someone needs a feature, they'll open a ticket and explain why they need it. Good software is always driven by needs and requests &ndash; feature set and code as short as can be.

Likewise about assertions, only assert something if it's a case that would subtly corrupt the output and be hard to debug or even realize it's wrong. And it oughta be a simple thing. e.g. I assert stack is empty when doing a line otherwise the state machine would accept offcurve followed by oncurve which is an instant GTFO

File: src/trufont/util/drawing.py (hunk 9 above)
Comment (adrientetar) [reply src/trufont/util/drawing.py]: ok I see the commit is named "hide handles". why hide? can you add a picture of what you want to do with handles?

File: src/trufont/objects/factories.py (hunk 4 above)
Comment (madig) [reply src/trufont/objects/factories.py]: Oops. Can't remember why I deleted this.

File: src/trufont/util/drawing.py (hunk 8 above)
Comment (madig) [reply src/trufont/util/drawing.py]: It's a cosmetic thing mostly, I find drawn handles on qcurve off-curves confusing because they suggest curve semantics. Leave in instead? Do something else?

File: src/trufont/objects/factories.py (hunk 5 above)
Comment (madig) [reply src/trufont/objects/factories.py]: It's part of the all-off-curve handling. Remove completely?

File: src/trufont/util/drawing.py (hunk 8 above)
Comment (adrientetar) [reply src/trufont/util/drawing.py]: it's not curve semantics, it's smooth point semantics

Comment (adrientetar) [reply src/trufont/util/drawing.py]: I would leave as-is personally

File: src/trufont/objects/factories.py (hunk 4 above)
Comment (adrientetar) [reply src/trufont/objects/factories.py]: for the "all off curves" thing probably.

File: src/trufont/objects/factories.py (hunk 3 above)
Comment (madig) [reply src/trufont/objects/factories.py]: Hm. Guess it can be turned into a `assert not stack` after the point loop?

File: src/trufont/objects/factories.py (hunk 6 above)
Comment (adrientetar) [reply src/trufont/objects/factories.py]: You only need to assert this in the else below

**Comment by madig** [comment]
The last assert doesn't seem to be needed actually as the other asserts catch all-off-curves already.

File: src/trufont/objects/factories.py [hunk 10]
```diff
@@ -97,6 +97,20 @@ def pathGraphicsPathFactory(path):
             if point.type == "curve":
                 graphicsPath.AddCurveToPoint(*stack)
                 stack = []
+            elif point.type == "qcurve":
+                if len(stack) == 4:
+                    graphicsPath.AddQuadCurveToPoint(*stack)
+                else:
+                    # Group off-curves into quads to construct implied on-curve points.
+                    assert len(stack) % 2 == 0
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: this can be removed, elements are always added in pairs so this is always true

File: src/trufont/objects/factories.py [hunk 11]
```diff
@@ -97,6 +97,20 @@ def pathGraphicsPathFactory(path):
             if point.type == "curve":
                 graphicsPath.AddCurveToPoint(*stack)
                 stack = []
+            elif point.type == "qcurve":
+                if len(stack) == 4:
+                    graphicsPath.AddQuadCurveToPoint(*stack)
+                else:
+                    # Group off-curves into quads to construct implied on-curve points.
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: "Group off-curves into quads" it's more like splitting than grouping, but I would say "make simple quads by constructing implied oc" because that's what's happening. large segment broken down to simple ones

File: src/trufont/objects/factories.py [hunk 12]
```diff
@@ -97,6 +97,20 @@ def pathGraphicsPathFactory(path):
...
+                    assert len(stack) % 2 == 0
+                    for offset in range(0, len(stack) - 4, 2):
+                        cx1, cy1, cx2, cy2 = stack[offset : offset + 4]
+                        graphicsPath.AddQuadCurveToPoint(
+                            cx1, cy1, (cx1 + cx2) / 2, (cy1 + cy2) / 2
+                        )
+                    # Use last two points as actual control and on-curve point.
```
Comment (adrientetar) [inline src/trufont/objects/factories.py]: can be removed, it's obvious from the code

File: src/trufont/objects/factories.py (hunk 11 above)
Comment (madig) [reply src/trufont/objects/factories.py]: I don't understand that sentence 😬

Comment (adrientetar) [reply src/trufont/objects/factories.py]: By constructing implied oc (`(cx1 + cx2) / 2, (cy1 + cy2) / 2`) we can make simple quads (a segment made of just one off followed by on which is the canonical form)

**Comment by adrientetar** [comment]
Btw, why do you need this? you wouldn't draw with quads?

File: src/trufont/objects/factories.py [hunk 13]
```diff
@@ -97,6 +97,19 @@ def pathGraphicsPathFactory(path):
             if point.type == "curve":
                 graphicsPath.AddCurveToPoint(*stack)
                 stack = []
+            elif point.type == "qcurve":
+                if len(stack) == 4:
+                    graphicsPath.AddQuadCurveToPoint(*stack)
+                else:
+                    # Iterate over off-curve points in quads (current, next) to
```
[the remark this answers is not in the document]
Comment (madig) [reply src/trufont/objects/factories.py]: Then away the comment goes.

**Comment by madig** [comment]
No, but I occasionally handle sources with converted curves and inspect TTFs.

File: src/trufont/objects/factories.py (hunk 11 above)
Comment (madig) [reply src/trufont/objects/factories.py]: Oh, ok. I thought by "quad" you meant the four elements in the stack slice.

**Comment by adrientetar** [comment]
(This still needs supporting changes to tfont, to be clear)

**Comment by adrientetar** [comment]
??
Why did you merge this without the supporting changes?

**Comment by madig** [comment]
Uh, well, it allows me to more easily test stuff as there's one less branch to juggle. Revert?

**Comment by adrientetar** [comment]
No need at this time but I mean let's not merge half-baked things unless we need to.

**Comment by madig** [comment]
Ok, sorry.

**Comment by adrientetar** [comment]
np, and thanks for the PR and for addressing all the review comments 🙂

**Comment by madig** [comment]
I just found a real-world example for an all-off-curves glyph: https://github.com/typesupply/defcon/blob/master/Lib/defcon/test/testdata/TestFont.ufo/glyphs.public.background/A_.glif

**Comment by adrientetar** [comment]
> I just found a real-world example

No, you found defcon test suite.

**Comment by moyogo** [comment]
https://github.com/TypeNetwork/Amstelvar was drawn with quadratic curves.

**Comment by adrientetar** [comment]
Are there tools that have a quadratic pen tool?

**Comment by jenskutilek** [comment]
If you want to work with TrueType hinting at some point, you should take extreme care

1. not to change the point indexes of imported outlines
2. to support everything that the glyf table spec allows, including all-offcurve contours.

**Comment by madig** [comment]
I just found an all-off-curve qcurve in the wild 😲 

DejaVu Sans ExtraLight, in the bowls of `six` and `nine`.
